src/runtime/hexagon_remote/remote.cpp

```diff
@@ -44,16 +90,220 @@ int halide_do_task(void *user_context, halide_task_t f, int idx,
+void goto_work(work *owned_job) {
+    work myjob;
+    // You can work only if you get a lock on the work queue.
+    unsigned int tid = qurt_thread_get_id();
+    FARF(LOW, "HVX_TP: %d: goto_work: Trying to get a lock on the work queue\n", tid);
+
+    // ***********************
+    // *** Lock work queue ***
+    // ***********************
+    lock_work_queue();
+    FARF(LOW, "HVX_TP: %d: goto_work: got a lock on the work queue\n", tid);
+
+    // If I'm a job owner, then I was the thread that called
+    // do_par_for, and I should only stay in this function while there
+    // are active workers on the job (i.e. running()) or if the job
+    // hasn't been claimed entirely (!claimed()). If I'm a lowly worker
+    // thread, I should stay in this function as long as the work queue is running.
+    while (owned_job != NULL ? (owned_job->running() || !owned_job->claimed())
+           : work_queue.running()) {
+
+        FARF(LOW, "HVX_TP: %d: goto_work: In the main goto_work loop \n", tid);
+        work *jobs = work_queue.jobs;
+        work *job;
+        if (!owned_job) {
+            // If this threads doesn't own a job, it looks for one
+            // and tries to do it. If it cannot find a job, it goes
+            // to sleep.
+            job = work_queue.jobs;
+            if (!job) {
+                // release the lock and go to sleep.
+                // ***************************
+                // *** Work queue unlocked ***
+                // ***************************
+                FARF(LOW, "HVX_TP: %d: goto_work: Couldn't find a job, going to sleep\n", tid);
+                unlock_work_queue();
+                qurt_sem_down(&wait_for_work);
+                FARF(LOW, "HVX_TP: %d: goto_work: Got woken up. Going to look for work.\n", tid);
+                 // ***********************
+                // *** Lock work queue ***
+                // ***********************
+                lock_work_queue();
+                continue;
+            } else {
+                myjob = *job;
+                FARF(LOW, "HVX_TP: %d: goto_work: Found a job working on x = %d\n", tid, myjob.next);
+            }
+        } else {
+            // We are here only if this thread owns a job that is not done.
+            // So do a part of it.
+            if (owned_job->claimed()) {
+                // the owner goes to sleep till the workers wake it up.
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner about to sleep\n", tid);
+                qurt_cond_wait(&owned_job->wakeup_owner, &work_queue.work_mutex);
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner waking up.\n", tid);
+                // This thread should have the lock now after having been woken up.
+                break;
+            } else {
+                myjob = *owned_job;
+                job = owned_job;
+                FARF(LOW, "HVX_TP: %d: goto_work: Owner about to work\n", tid);
+            }
+        }
+        job->next++;
```

Comment:
I think you don't need the local copy of myjob if you just say 'int idx = job->next++', and then use idx below (instead of myjob.next). All the rest of the references to myjob can just be through the job pointer.